packages/wp-build/src/build.mjs

```diff
@@ -964,14 +964,78 @@ async function transpilePackage( packageName ) {
 		);
 	}
 
-	const srcFiles = await glob(
-		normalizePath(
-			path.join( packageDir, `src/**/*.${ SOURCE_EXTENSIONS }` )
-		),
-		{
-			ignore: IGNORE_PATTERNS,
+	const cjsEntryPoints = new Set();
+	const moduleEntryPoints = new Set();
+
+	function addBuildPath( entryPoints, buildPath ) {
+		// Convert an exported path like `./build-module/something.js` to
+		// the corresponding source path like `./src/something.tsx`.
+		const srcPath = buildPath
+			.replace( /\/build-module\/|\/build\//, '/src/' )
+			.replace( /\.[^.]+$/, `.${ SOURCE_EXTENSIONS }` );
+		const srcPathNormalized = normalizePath(
+			path.join( packageDir, srcPath )
+		);
+		// Find the real source file, with one of the supported extensions, with `glob`.
+		const srcFiles = glob.sync( srcPathNormalized );
+		if ( srcFiles.length > 0 ) {
+			entryPoints.add( srcFiles[ 0 ] );
 		}
-	);
+	}
+
+	// Add the `.` exports as entrypoints, both CJS and ESM.
+	if ( packageJson.exports ) {
+		const rootExport = packageJson.exports[ '.' ];
+		if ( typeof rootExport === 'string' ) {
+			addBuildPath( cjsEntryPoints, rootExport );
+			addBuildPath( moduleEntryPoints, rootExport );
+		} else if ( typeof rootExport === 'object' ) {
+			const rootCjsExport = rootExport.require || rootExport.default;
+			if ( rootCjsExport ) {
+				addBuildPath( cjsEntryPoints, rootCjsExport );
+			}
+			const rootModuleExport = rootExport.import || rootExport.default;
+			if ( rootModuleExport ) {
+				addBuildPath( moduleEntryPoints, rootModuleExport );
+			}
+		}
+	}
+
+	// Add `index` and `init` scripts for each block.
+	if ( packageName === 'block-library' ) {
```

Comment:
Wonder if there's a way we can make this more generic without building in awareness of specific packages. I think the approach has generally been to try to abstract custom `wp*` properties that are defined at the package level to support specific behaviors like this.